iiod/iiod.c

```diff
@@ -106,57 +110,255 @@ static const char *options_descriptions[] = {
+			create_services(avahi_entry_group_get_client(group));
+			break;
+		}
+		case AVAHI_ENTRY_GROUP_FAILURE :
+			ERROR("Entry group failure: %s\n",
+					avahi_strerror(avahi_client_errno(
+						avahi_entry_group_get_client(group))));
+			break;
+		case AVAHI_ENTRY_GROUP_UNCOMMITED:
+			/* This is normal,
+			 * since we commit things in the create_services()
+			 */
+			DEBUG("Avahi: Group uncommited\n");
+			break;
+		case AVAHI_ENTRY_GROUP_REGISTERING:
+			DEBUG("Avahi: Group registering\n");
+			break;
+	}
 }
 
 static void __avahi_client_cb(AvahiClient *client,
 		AvahiClientState state, void *d)
 {
-	AvahiEntryGroup *group;
-
-	if (state != AVAHI_CLIENT_S_RUNNING)
+	if (!client) {
+		ERROR("__avahi_client_cb with no valid client\n");
 		return;
+	}
 
-	group = avahi_entry_group_new(client, __avahi_group_cb, NULL);
-
-	if (group && !avahi_entry_group_add_service(group,
-			AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC,
-			0, "iio", "_iio._tcp", NULL, NULL, IIOD_PORT, NULL)) {
-		avahi_entry_group_commit(group);
-		INFO("Registered to ZeroConf server %s\n",
-				avahi_client_get_version_string(client));
+	switch (state) {
+		case AVAHI_CLIENT_S_RUNNING:
+			/* Same as AVAHI_SERVER_RUNNING */
+			DEBUG("Avahi: create services\n");
+			/* The server has startup successfully, so create our services */
+			create_services(client);
+			break;
+		case AVAHI_CLIENT_FAILURE:
+			ERROR("Avahi: Client failure: %s\n",
+					avahi_strerror(avahi_client_errno(client)));
+			client_free();
+			avahi.client = client_new();
+			break;
+		case AVAHI_CLIENT_S_COLLISION:
+			/* Same as AVAHI_SERVER_COLLISION */
+			/* When the server is back in AVAHI_SERVER_RUNNING state
+			 * we will register them again with the new host name. */
+			DEBUG("Avahi: Client collision\n");
+			/* Let's drop our registered services.*/
+			if (avahi.group)
+				avahi_entry_group_reset(avahi.group);
+			break;
+		case AVAHI_CLIENT_S_REGISTERING:
+			/* Same as AVAHI_SERVER_REGISTERING */
+			DEBUG("Avahi: Client group reset\n");
+			if (avahi.group)
+				avahi_entry_group_reset(avahi.group);
+			break;
+		case AVAHI_CLIENT_CONNECTING:
+			DEBUG("Avahi: Client Connecting\n");
+			break;
 	}
 
 	/* NOTE: group is freed by avahi_client_free */
 }
 
+static AvahiClient * client_new(void)
+{
+	int ret;
+	AvahiClient * client;
+
+	client = avahi_client_new(avahi_threaded_poll_get(avahi.poll),
+			0, __avahi_client_cb, NULL, &ret);
+	if (!client) {
+		ERROR("Avahi: failure creating client: %s (%d)\n",
+				avahi_strerror(ret), ret);
+	}
+
+	return client;
+}
+
+
+static void create_services(AvahiClient *c)
+{
+	int ret;
+
+	if (!c) {
+		ERROR("create_services called with no valid client\n");
+		goto fail;
+	}
+
+	if (!avahi.group) {
+		avahi.group = avahi_entry_group_new(c, __avahi_group_cb, NULL);
+		if (!avahi.group) {
+			ERROR("avahi_entry_group_new() failed: %s\n",
+					avahi_strerror(avahi_client_errno(c)));
+			goto fail;
+		}
+	}
+
+	if (avahi_entry_group_is_empty(avahi.group)) {
```

Comment:
this can be left as-is;
but this can save one level of indentation [for this if-block] by doing:
```
if (!avahi_entry_group_is_empty(avahi.group)) {
    DEBUG("Avahi group not empty\n");
     return;
}
```

Reply:
thanks

Reply:
should be updated in latest pull request.